pkg/analytics/instrumentation_collector.go

```diff
@@ -0,0 +1,242 @@
+	ic.instrumentationErr = append(ic.instrumentationErr, err)
+}
+
+func (ic *instrumentationCollectorImpl) AddExtension(key string, value interface{}) {
+	if ic.extension == nil {
+		ic.extension = make(map[string]interface{})
+	}
+	ic.extension[key] = value
+}
+
+func GetV2InstrumentationObject(collector InstrumentationCollector) (*api.AnalyticsRequestBody, error) {
+	t, ok := collector.(*instrumentationCollectorImpl)
+	if ok {
+		return t.GetV2InstrumentationObject()
+	}
+	return nil, fmt.Errorf("failed to convert collector")
+}
+
+func (ic *instrumentationCollectorImpl) GetV2InstrumentationObject() (*api.AnalyticsRequestBody, error) {
+	a, err := ic.getV2Attributes()
+
+	if err != nil {
+		return nil, err
+	}
+
+	d := api.AnalyticsData{
+		Type:       ic.instrumentationType,
+		Attributes: a,
+	}
+
+	return &api.AnalyticsRequestBody{
+		Data: d,
+	}, nil
+}
+
+func (ic *instrumentationCollectorImpl) getV2Attributes() (api.AnalyticsAttributes, error) {
+	r, err := ic.getV2Runtime()
+
+	if err != nil {
+		return api.AnalyticsAttributes{}, err
+	}
+
+	return api.AnalyticsAttributes{
+		Interaction: ic.getV2Interaction(),
+		Runtime:     r,
+	}, nil
+}
+
+func (ic *instrumentationCollectorImpl) getV2Interaction() api.Interaction {
+	return api.Interaction{
+		Categories:  &ic.category,
+		Errors:      toInteractionErrors(ic.instrumentationErr),
+		Extension:   &ic.extension,
+		Id:          ic.interactionId,
+		Results:     toInteractionResults(&ic.testSummary),
+		Stage:       toInteractionStage(&ic.stage),
+		Status:      string(ic.status),
+		Target:      api.Target{Id: ic.targetId},
+		TimestampMs: ic.timestamp.UnixMilli(),
+		Type:        ic.instrumentationType,
+	}
+}
+
+func (ic *instrumentationCollectorImpl) getV2Runtime() (*api.Runtime, error) {
+	hasUserAgentAppData := len(ic.userAgent.App) > 0 || len(ic.userAgent.AppVersion) > 0
+	if !hasUserAgentAppData {
+		return nil, errors.New("no user agent application data")
+	}
+
+	return &api.Runtime{
+		Application: &api.Application{
+			Name:    ic.userAgent.App,
+			Version: ic.userAgent.AppVersion,
+		},
+		Environment: &api.Environment{
+			Name:    ic.userAgent.IntegrationEnvironment,
+			Version: ic.userAgent.IntegrationEnvironmentVersion,
+		},
+		Integration: &api.Integration{
+			Name:    ic.userAgent.Integration,
+			Version: ic.userAgent.IntegrationVersion,
+		},
+		Performance: &api.Performance{
+			DurationMs: ic.duration.Milliseconds(),
+		},
+		Platform: &api.Platform{
+			Arch: ic.userAgent.Arch,
+			Os:   ic.userAgent.OS,
+		},
+	}, nil
+}
+
+func toInteractionResults(testSummary *json_schemas.TestSummary) *[]map[string]interface{} {
+	r := []map[string]interface{}{}
+	for _, result := range testSummary.Results {
+		r = append(r, map[string]interface{}{
+			"name":  result.Severity,
+			"count": result.Total,
+		})
+	}
+	return &r
+}
+
+func toInteractionStage(s *string) *api.InteractionStage {
+	var is api.InteractionStage
+
+	switch stage := s; *stage {
```

Comment:
Suggestion: this looks like it can be made easier. Overall InteractionStage is a string, so a type cast would be sufficient. I don't see much value of enforcing correct strings here, if there are values not acceptable, the api should handle this. Otherwise we are spreading knowledge unnecessarily.

Reply:
hmm, I'm not sure. We have the [schema](https://github.com/snyk/go-application-framework/blob/661cc24c0edb47aabade15fb050b07168555610f/internal/api/clients/analytics_v20240307.go#L20-L26) which defines the valid values for this property.

IMO, we should do our best to conform to the schema contract, and so where a schema defines valid values, we should  add checks to make sure this is the case.

The alternative is we only let the API handle this validation, which will cost an API call.